static/js/containers/DashboardPage_test.js

```diff
@@ -326,4 +342,82 @@ describe('DashboardPage', () => {
       });
     });
   });
+
+  describe('handles redeeming coupons', () => {
+    const COUPON_SUCCESS_ACTIONS = DASHBOARD_SUCCESS_ACTIONS.concat([
+      REQUEST_ATTACH_COUPON,
+      RECEIVE_ATTACH_COUPON_SUCCESS,
+      SET_RECENTLY_ATTACHED_COUPON,
+      SET_COUPON_NOTIFICATION_VISIBILITY,
+      REQUEST_FETCH_COUPONS,
+      RECEIVE_FETCH_COUPONS_SUCCESS,
+    ]);
+    const COUPON_FAILURE_ACTIONS = DASHBOARD_SUCCESS_ACTIONS.concat([
+      REQUEST_ATTACH_COUPON,
+      RECEIVE_ATTACH_COUPON_FAILURE,
+      SET_TOAST_MESSAGE,
+    ]);
+
+    it('with a successful fetch', () => {
+      helper.couponsStub.returns(Promise.resolve([COUPON]));
+
+      return renderComponent(
+        '/dashboard?coupon=success-coupon',
+        COUPON_SUCCESS_ACTIONS
+      ).then(() => {
+        const state = helper.store.getState();
+        assert.deepEqual(state.coupons.recentlyAttachedCoupon, COUPON);
+        assert.isTrue(state.ui.couponNotificationVisibility);
+        assert.deepEqual(state.coupons.coupons, [COUPON]);
+      });
+    });
+
+    it('with a failed fetch', () => {
+      helper.attachCouponStub.returns(Promise.reject());
+
+      return renderComponent(
+        '/dashboard?coupon=failure-coupon',
+        COUPON_FAILURE_ACTIONS
+      ).then(() => {
+        const state = helper.store.getState();
+        assert.isNull(state.coupons.recentlyAttachedCoupon);
+        assert.isFalse(state.ui.couponNotificationVisibility);
+        assert.deepEqual(state.ui.toastMessage, {
+          title: "Coupon failed",
+          message: "This coupon code is invalid or does not exist.",
+          icon: TOAST_FAILURE
+        });
+      });
+    });
+
+    // This test is disabled because I can't figure out a good way to simulate
+    // this race condition in an automated test. If someone else can pick it
+    // up and fix it, that would be great!
+    xit('without a race condition', () => {  // eslint-disable-line mocha/no-skipped-tests
```

Comment:
When I run this test, it says that it's not receiving the `RECEIVE_GET_PROGRAM_ENROLLMENTS_SUCCESS` action, which is odd -- I don't know why the `getPrograms()` API would be unable to return if the `getCoupons()` API is slow. More concerning is that the `REQUEST_ATTACH_COUPON` action and `RECEIVE_ATTACH_COUPON_SUCCESS` actions apparently _are_ being received in this test -- I was trying to write it in such a way that the code would return early so that `attachCoupon()` would never get called, but apparently it's still getting called. (As an aside, is there an easy way to assert that a particular action is _not_ called with `renderComponent`?)

If we can find a way to write this test correctly, so that it asserts on the right thing and passes reliably, then we should do so. Otherwise, I think we should just delete this test and leave the race condition avoidance code untested. @noisecapella: what do you think? Can you see how to fix this?